src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs

```diff
@@ -1276,5 +1297,203 @@ private bool TryTranslateSumResultOperator(PipelineExpression node, out BsonValu
             result = null;
             return false;
         }
+
+        internal class TemplatesState
```

Comment:
I guess this class requires a number of unit tests. I will add them if the current solution is ok.

Reply:
Yeah... the unit tests will show me what is actually going on in here. I think I get it, but they will clarify.

Reply:
I will add it tomorrow. I will let you know.